src/libs/antares/study/scenario-builder/rules.cpp

```diff
@@ -361,7 +379,8 @@ bool Rules::apply()
             returned_status = renewable[i].apply(study_) && returned_status;
             returned_status = linksNTC[i].apply(study_) && returned_status;
         }
-        returned_status = hydroLevels.apply(study_) && returned_status;
+        returned_status = hydroInitialLevels.applyHydroLevels(study_.scenarioInitialHydroLevels) && returned_status;
```

Comment:
About : 
```C++
    returned_status = hydroInitialLevels.applyHydroLevels(study_.scenarioInitialHydroLevels) && returned_status;
    returned_status = hydroFinalLevels.applyHydroLevels(study_.scenarioFinalHydroLevels) && returned_status;
```
We're aware that the TS numbers data classes hierarchy have shortcomings, and maybe that's the reason why you chose not to implement a proper **apply(study)** for hydro levels (which is a pure virtual method).
Until we address these shortcomings, we should stick to the plan, implement and call it here : 
```C++
    returned_status = hydroInitialLevels.apply(study_) && returned_status;
    returned_status = hydroFinalLevels.apply(study_) && returned_status;
```

On way to do this would be to use dependency injection : 
```C++
// hydroLevelsData.h
class hydroLevelsData final : public dataInterface
{
    ....
 private:
    std::function<void(Study&, MatrixType&)> applyToTarget_;
    ....
}; // class hydroLevelsData

void initLevelApply(Study& study, MatrixType& matrix)
{
     study.scenarioInitialHydroLevels.copyFrom(matrix);
}

void finalLevelApply(Study& study, MatrixType& matrix)
{
     study.scenarioFinalHydroLevels.copyFrom(matrix);
}
...
```
```C++
// hydroLevelsData.cpp
...
hydroLevelsData::hydroLevelsData(std::string& iniFilePrefix, 
                                 std::function<void(Study&, MatrixType&)> applyToTarget) :
    addToPrefix_(iniFilePrefix),
    applyToTarget_(applyToTarget)
{}
...

bool hydroLevelsData::apply(Study& study)
{
    applyToTarget_(study, pHydroLevelsRules);
    return true;
}
```
```C++
// rules.cpp
...

Rules::Rules(Study& study) : study_(study), pAreaCount(0)
{
    std::string prefixInitial = "hl,";
    hydroInitialLevels = hydroLevelsData(prefixInitial, initLevelApply);

    std::string prefixFinal = "hfl,";
    hydroFinalLevels = hydroLevelsData(prefixFinal, initLevelApply);
}
```

Reply:
done